### ISSUE TITLE
[BUG]: AutoHiddenGroup elements do not properly process AppearanceChanged events

### ISSUE THREAD

**Issue by MGRussell** [Wagnerp/Krypton-Toolkit-Suite-NET-Core#30]
**Issue Description**

This issue was discovered working on #29 .

When you change appearance elements of an `AutoHiddenGroup` item in a `KryptonDockableWorkspace`, that change is not visually reflected. This seems to be caused by issues relating to the way that the `AutoHiddenGroup` uses the real `KryptonPage`s and then the `KryptonAutoHiddenProxyPage`s. The event ends up getting wired to the proxy page instead of the real page, and is never called.

I've created a project to demonstrate this issue [here](https://github.com/MGRussell/KryptonAutoHiddenGroupIssue).

---

**Sample Project Description**

This is a screenshot of the UI:

![autohiddengroupissue](https://user-images.githubusercontent.com/30259368/74512847-be254600-4f3b-11ea-95ba-ae36ef6e33a8.PNG)

The UI has setup two pages. One is a normal workspace page - Page 1. The other is an `AutoHiddenGroup` page - Page 2. The three toggle icons below change the page's `KryptonPage.ImageSmall` image as follows:

- Toggle Regular: This toggles the image for Page 1. 

- Toggle Hidden: This toggles the image for Page 2.

- Toggle Hidden Proxy: This toggles the image for the internal `KryptonAutoHiddenProxyPage` element representing Page 2.

Finally, the "Force OnAppearanceChanged" determines whether or not the specified page is forced to call OnAppearancePropertyChanged (achieved through reflection). 

---

**Expected/Observed behavior:**

- Toggle Regular: This should toggle the image for Page 1. This works 100% correctly.

- Toggle Hidden: This should toggle the image for Page 2. This half works. If you toggle the image, you can then see it rendered if you manually force the Page 2 tab to redraw by, for instance, mousing over it. However, you will see nothing unless you force a manual redraw.

- Toggle Hidden Proxy: This should not do anything. However, if you force the OnAppearancePropertyChanged property to fire it acts exactly like the "Toggle Hidden" button *should* work, and it changes with image for Page 2 with an instantaneous visual update.

---

**Solution:**

 It seems that this issue should be able to be resolved by ensuring that the `OnAppearanceChanged` callback from the KryptonNavigator is hooked into the `OnApperanceChanged` of the real `KryptonPage` and not the `KryptonAutoHiddenProxyPage`. However, this seems to be easier said than done in going over the code.

---

**The following video demonstrates the issue:**

![eOEdEfbPeL](https://user-images.githubusercontent.com/30259368/74515018-4f96b700-4f40-11ea-83cd-ef4a11538a33.gif)

**Text of what I've written in the video in case I did not allot sufficient time:**

---

First, the regular tab. Everything works as expected here.

---

Now for the AutoHiddenGroup tab. Here the update only works if manually triggered.
Forcing OnApperancePropertyChanged to fire has no effect, because the event is null.

---

Now for the proxy of the AutoHiddenGroup tab. Here it works similarly to the AutoHiddenGroup tab.

However, forcing OnApperancePropertyChanged makes everything work as it should for the regular AutoHiddenGroup page. This is because the visual refresh callback is connected to the KryptonAutoHiddenProxyPage's event instead of the 'real' KryptonPage's.

**Comment by Wagnerp** [comment]
Hi @MGRussell 

Thanks for posting your issue, the problem that I have on my end is broken links

![image](https://user-images.githubusercontent.com/949607/74552984-49a7d280-4f4e-11ea-9441-2837723ebf10.png)

So sorry for the delay in responses.

**Comment by MGRussell** [comment]
How absolutely peculiar. I'm hosting the images on github and can see them in multiple browsers so it should not be because of caching or privileges or anything like that. Only thing I can imagine is maybe there was some sort of a temporary CDN issue with github? Can you let me know if you still can't see them? I'll try reuploading them if you cannot.

Much appreciated. And of course no problem at all on the time frames. We're both playing the tortoise. :1st_place_medal:

**Comment by MGRussell** [comment]
OK after looking through the code more, I can sum this issue up much more precisely now - and also now have a passable solution:

----

**Problem**

Adding a `KryptonPage` to an `AutoHiddenGroup` in a `DockableWorkSpace` results in that page being added to an underlying `KryptonNavigator`.

The page sent to the `KryptonNavigator` is a `KryptonAutoHiddenProxyPage` instead of the actual `KryptonPage`. This is intentional and by design. Because of this, when the `KryptonNavigator` adds its callbacks for various events (see: KryptonNavigator.cs line 2410), including `OnAppearanceChanged`, it adds them to the proxy version of the page.

The problem this causes is that the proxy page is 'dumb'. When you do something like change the image on a `KryptonAutoHiddenProxyPage` it does nothing except pass the call on to its underlying `KryptonPage`:

        // KryptonAutoHiddenProxyPage.cs
        public override Image ImageSmall
        {
            get => Page.ImageSmall;
            set => Page.ImageSmall = value;
        }

And there is the issue! The underlying Page now changes its image and fires its  `OnAppearanceChanged` event as follows:

        // KryptonPage.cs
        public virtual Image ImageSmall
        {
            [System.Diagnostics.DebuggerStepThrough]
            get { return _imageSmall; }

            set
            {
                if (_imageSmall != value)
                {
                    _imageSmall = value;
                    OnAppearancePropertyChanged("ImageSmall");
                }
            }
        }

But like mentioned, the AppearancePropertyChanged event is hooked up on the `KryptonAutoHiddenProxyPage` and not the real `KryptonPage`! So changing the image (using either of the proxy or real page) results in no callback being fired, and no visual update.

-----

**Solution**

As a 'proof of concept' type solution, I made the KryptonPage OnAppearanceChanged event virtual. And in KryptonAutoHiddenProxyPage I overrode it as:

        public override event PropertyChangedEventHandler AppearancePropertyChanged
        {
            add 
            { 
                if(Page != null)
                    Page.AppearancePropertyChanged += value;
            }
            remove 
            {
                if(Page != null)
                    Page.AppearancePropertyChanged -= value;
            }
        }

And now everything works perfectly. The reason I call this a proof of concept solution is because while it solves this specific issue completely, it is possible that there are other properties facing the exact same problem due to the incomplete interlinking between the proxy page and the real page. It seems there should be some way to resolve all these problems at once.

**Comment by Wagnerp** [comment]
Hi @MGRussell 

Yes same issue, if I click on one of the images, I get `ERR_SSL_VERSION_OR_CIPHER_MISMATCH`. Can't pinpoint the issue, had to spin up a old Vista VM to see the assets!

So, is the problem in `KryptonAutoHiddenProxyPage.cs`?

**Comment by MGRussell** [comment]
Yes, the `KryptonAutoHiddenProxyPage` is supposed to work as a proxy/pass-through class to the `KryptonPage` class, yet subscribed events such as `AppearancePropertyChanged` can end up getting "stranded" in the proxy class in certain scenarios.

Normal usage of an `AutoHiddenGroup` is one of those scenarios. When pages are added an `AutoHiddenGroup` the underlying `KryptonNavigator` is sent the proxy page instead of the real page. This is intentional.

However, this causes the navigator to then subscribe to the proxy events. And so when you change something on the real page that would trigger an `AppearancePropertyChanged` event, such as changing the `ImageSmall` property, the `AppearancePropertyChanged` event is never propagated to the `KryptonNavigator`, and thus it is not reflected in the UI.

**Comment by Wagnerp** [comment]
Ok, I'll add in the code to be rolled out in build 2003

**Comment by Wagnerp** [comment]
Ok, I've made the edits, so it'll be in build 2003

**Comment by MGRussell** [comment]
Great, much appreciated!
